### PR TITLE
fix: Fixes focus controls in drawers

### PR DESCRIPTION
**Pull request by katiegeorge**
### Description

Part of Drawers Bug Bash.

This simplifies and fixes the drawer focus controls to use the same hook as tools, and also adds a test to make sure focus works as expected with overflow and on-page triggers.

Related links, issue #, if available: n/a

### How has this been tested?

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/app-layout/drawer/index.tsx [hunk 1]
```diff
@@ -182,12 +182,17 @@ export const DrawerTriggersBar = ({
...
   drawers,
+  drawerRefs,
   onDrawerChange,
 }: DrawerTriggersBarProps) => {
   const containerRef = React.useRef<HTMLDivElement>(null);
+  const previousActiveDrawerId = useRef(activeDrawerId);
   const [containerHeight, triggersContainerRef] = useContainerQuery(rect => rect.contentBoxHeight);
   const isCompactMode = useDensityMode(containerRef) === 'compact';
 
+  if (activeDrawerId) {
```
Comment (avinashbot) [inline src/app-layout/drawer/index.tsx:193]: I'm not sure what this code is doing, could you explain it? Why assign a ref only if the ref was previously active?

File: src/app-layout/interfaces.ts [hunk 2]
```diff
@@ -220,6 +220,11 @@ export namespace AppLayoutProps {
      */
     focusToolsClose(): void;
 
+    /**
+     * Focuses the active drawer. Use this to focus the active drawer after opening it programmatically.
+     */
+    focusActiveDrawerClose(): void;
```
Comment (avinashbot) [inline src/app-layout/interfaces.ts:226]: Non-blocking, but I guess the name's not really accurate anymore, is it? For resizable drawers, we focus on the slider, so maybe it should be closer to split panel's `focusActiveDrawer`? There was a fair bit of discussion about naming when it came to split panel too so... I'll leave it up to you.

Comment (katiegeorge) [reply src/app-layout/interfaces.ts]: Good point, I updated it to just `focusActiveDrawer`.

File: src/app-layout/drawer/index.tsx (hunk 1 above)
Comment (katiegeorge) [reply src/app-layout/drawer/index.tsx:193]: We do the same thing [here](https://github.com/cloudscape-design/components/blob/main/src/app-layout/visual-refresh/drawers.tsx#L184), with the comment: 
> Changes to the activeDrawerId need to be tracked by the previousActiveDrawerId property in order to appropriately apply the ref required to manage focus control.

Comment (katiegeorge) [reply src/app-layout/drawer/index.tsx:193]: I think its basically to make sure the focus control works property on drawer change, not just open / close. When I just checked, I realized I did have to add a new dependency to the focus hook for when the active id changes

File: src/app-layout/utils/use-focus-control.ts [hunk 3]
```diff
@@ -41,9 +51,8 @@ export function useFocusControl(isOpen: boolean, restoreFocus = false): FocusCon
     shouldFocus.current = false;
   };
 
-  // We explictly want this effect to run when only `isOpen` changes
   // eslint-disable-next-line react-hooks/exhaustive-deps
-  useEffect(doFocus, [isOpen]);
+  useEffect(doFocus, [isOpen, dependencies]);
```
Comment (avinashbot) [inline src/app-layout/utils/use-focus-control.ts:55]: You probably want this, or else it'll fire on every render because the array is a new object each time:

```suggestion
  useEffect(doFocus, [isOpen, ...dependencies]);
```

But passing dependencies through hooks isn't ideal, and React recommends against it. Maybe it's a sign that we shouldn't be doing useEffect in here and instead have the calling component call `setFocus` when its own dependencies change.

Comment (just-boris) [reply src/app-layout/utils/use-focus-control.ts:55]: This code is a generic support for "when switching from non-resizable drawer to resizable the resize handle should be focused"

Do you have any suggestions about this high level behavior? Where should the focus go when switching between already open drawers?

Comment (just-boris) [reply src/app-layout/utils/use-focus-control.ts:55]: +1 to that comment. I think we can avoid changing this effect all together. Let's do programmatic focus inside drawer toggle click handler

File: src/app-layout/drawer/index.tsx (hunk 1 above)
Comment (just-boris) [reply src/app-layout/drawer/index.tsx:193]: Looks like this code is intended to do the same job as this already existing code: https://github.com/cloudscape-design/components/blob/e9fbf2367717b27300bc11c194ca66a60b3f0af2/src/app-layout/utils/use-focus-control.ts#L34-L37

Can we leverage the existing `loseFocus` feature somehow?

File: src/app-layout/utils/use-focus-control.ts (hunk 3 above)
Comment (katiegeorge) [reply src/app-layout/utils/use-focus-control.ts:55]: What do you mean by this?

Comment (just-boris) [reply src/app-layout/utils/use-focus-control.ts:55]: We can keep `useFocusControl` unchanged and do `focusToolsButtons(true)` when we switch drawers

File: src/app-layout/drawer/index.tsx (hunk 1 above)
Comment (katiegeorge) [reply src/app-layout/drawer/index.tsx:193]: It's different, I believe. This section of the code only gets run when programmatically opening a drawer, not when changing drawers via triggers. Its working as expected now, let's keep it this way and we can revisit later if we need.

File: src/app-layout/utils/use-focus-control.ts (hunk 3 above)
Comment (katiegeorge) [reply src/app-layout/utils/use-focus-control.ts]: From DMs: explicitly passing down the `activeDrawerId` instead of dependencies here.

**Review by just-boris — APPROVED** [review]

**Review by avinashbot — APPROVED** [review]
